src/Microsoft.AspNetCore.Razor.LanguageServer/DefaultTagHelperCompletionService.cs

```diff
@@ -0,0 +1,299 @@
+            }
+
+            var parent = owner.Parent;
+            if (TryGetElementInfo(parent, out var containingTagNameToken, out var attributes) &&
+                containingTagNameToken.Span.IntersectsWith(location.AbsoluteIndex))
+            {
+                var stringifiedAttributes = StringifyAttributes(attributes);
+                var elementCompletions = GetElementCompletions(parent, containingTagNameToken.Content, stringifiedAttributes, codeDocument);
+                return elementCompletions;
+            }
+
+            // Invalid location for TagHelper completions.
+            return Array.Empty<CompletionItem>();
+        }
+
+        private IReadOnlyList<CompletionItem> GetElementCompletions(
+            SyntaxNode containingTag,
+            string containingTagName,
+            IEnumerable<KeyValuePair<string, string>> attributes,
+            RazorCodeDocument codeDocument)
+        {
+            var ancestors = containingTag.Ancestors();
+            var tagHelperDocumentContext = codeDocument.GetTagHelperContext();
+            var (ancestorTagName, ancestorIsTagHelper) = GetNearestAncestorTagInfo(ancestors);
+            var elementCompletionContext = new ElementCompletionContext(
+                tagHelperDocumentContext,
+                existingCompletions: Enumerable.Empty<string>(),
+                containingTagName,
+                attributes,
+                ancestorTagName,
+                ancestorIsTagHelper,
+                HtmlSchemaTagNames.Contains);
+
+            var completionItems = new List<CompletionItem>();
+            var completionResult = _razorTagHelperCompletionService.GetElementCompletions(elementCompletionContext);
+            foreach (var completion in completionResult.Completions)
+            {
+                var razorCompletionItem = new CompletionItem()
+                {
+                    Label = completion.Key,
+                    InsertText = completion.Key,
+                    FilterText = completion.Key,
+                    SortText = completion.Key,
+                    Kind = CompletionItemKind.TypeParameter,
+                    CommitCharacters = ElementCommitCharacters,
+                };
+
+                completionItems.Add(razorCompletionItem);
+            }
+
+            return completionItems;
+        }
+
+        // Internal for testing
+        internal static IEnumerable<KeyValuePair<string, string>> StringifyAttributes(SyntaxList<RazorSyntaxNode> attributes)
+        {
+            var stringifiedAttributes = new List<KeyValuePair<string, string>>();
+
+            for (var i = 0; i < attributes.Count; i++)
+            {
+                var attribute = attributes[i];
+                if (attribute is MarkupTagHelperAttributeSyntax tagHelperAttribute)
+                {
+                    var name = tagHelperAttribute.Name.GetContent();
+                    var value = tagHelperAttribute.Value.GetContent();
```

Comment:
Is Value guaranteed to not be null? (Probably more of a question to myself 😆). In any case, might be better to be defensive.